src/snowflake/connector/query_context_cache.py

```diff
@@ -0,0 +1,302 @@
+        # entry with id = 0 is the main entry
+        self._id = id
+        self._read_timestamp = read_timestamp
+        # priority values are 0..N with 0 being the highest priority
+        self._priority = priority
+        # OpaqueContext field will be base64 encoded in GS, but it is opaque to client side. Client side should not do decoding/encoding and just store the raw data.
+        self._context = context
+
+    @property
+    def id(self) -> int:
+        return self._id
+
+    @property
+    def read_timestamp(self) -> int:
+        return self._read_timestamp
+
+    @read_timestamp.setter
+    def read_timestamp(self, timestamp: int) -> None:
+        self._read_timestamp = timestamp
+
+    @property
+    def priority(self) -> int:
+        return self._priority
+
+    @property
+    def context(self) -> str:
+        return self._context
+
+    @context.setter
+    def context(self, ctx: str) -> None:
+        self._context = ctx
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, QueryContextElement):
+            return False
+        return (
+            self._id == other.id
+            and self._read_timestamp == other.read_timestamp
+            and self._priority == other.priority
+            and self._context == other.context
+        )
+
+    def __lt__(self, other: object) -> bool:
+        if not isinstance(other, QueryContextElement):
+            raise NotImplementedError(
+                f"cannot compare QueryContextElement with object of type {type(other)}"
+            )
+        return self._priority < other._priority
+
+    def __hash__(self) -> int:
+        _hash = 31
+
+        _hash = _hash * 31 + self._id
+        _hash += (_hash * 31) + self._read_timestamp
+        _hash += (_hash * 31) + self._priority
+        if self._context:
+            _hash += (_hash * 31) + int.from_bytes(
+                md5(self._context.encode("utf-8")).digest(), "big"
+            )
+        return _hash
+
+    def __str__(self) -> str:
+        return f"({self._id}, {self._read_timestamp}, {self._priority})"
+
+
+class QueryContextCache:
+    def __init__(self, capacity: int):
+        self._capacity = capacity
+        self._id_map: dict[int, QueryContextElement] = {}
+        self._priority_map: dict[int, QueryContextElement] = {}
+        self._intermediate_priority_map: dict[int, QueryContextElement] = {}
+
+        # stores elements sorted by priority. Element with
+        # least priority value has the highest priority
+        self._tree_set: set[QueryContextElement] = SortedSet()
+        self._lock = Lock()
+        self._data: str = None
+
+    @property
+    def capacity(self) -> int:
+        return self._capacity
+
+    def _add_qce(self, qce: QueryContextElement) -> None:
+        self._tree_set.add(qce)
+        self._id_map[qce.id] = copy.deepcopy(qce)
+        self._intermediate_priority_map[qce.priority] = copy.deepcopy(qce)
```

Comment:
@sfc-gh-lqin can you answer this one?

Reply:
@sfc-gh-mkeller Could you check this fix PR https://github.com/snowflakedb/snowflake-connector-python/pull/1532 ?  On JDBC side the implementation does not force a deepcopy. But when I was testing the code, I found a very tricky issue:

Basically, in my end-to-end test, when I want to

def _remove_qce(self, qce: QueryContextElement) -> None:
        self._id_map.pop(qce.id)
        self._priority_map.pop(qce.priority)
        self._tree_set.remove(qce)
even if there is an element in tree_set with all the same field values(id, ts, priority, context), the remove API can not find the entry to remove. I checked the hash values are the same. If i do

for ele in self._tree_set:
   if(ele.__eq__(qce)):
       self._tree_set.remove(ele)
This can work.

This reason could be that the python QueryContextElement object in

        self._id_map: dict[int, QueryContextElement] = {}
        self._priority_map: dict[int, QueryContextElement] = {}
        self._tree_set: set[QueryContextElement] = SortedSet()
are referencing the same underlying object. I use copy.deepcopy when adding a QCE to each map/set to avoid this tricky bug. In this way, we also can not do in-place update in merge function. The overhead should be very small since the QCC capacity is usually a small number.


@sfc-gh-mkeller If you can figure out other easier fix, please let me know and we can avoid such deepcopy.

Reply:
I looked into this issue separately and it seems like the whole problem with treeset was because we were doing an in-place update to query context elements which created problems. Lianke fixed this issue in an earlier PR but did not remove the deepcopy. 
I verified that the treeset problem was due to in-place updates and end-to-end tests works completely fine even without deepcopy. I have removed it in the latest commit.